### PR TITLE
1505-feature-chroma-support-metadata-filtering

### LINKED ISSUE #1505: [FEATURE] Chroma: support metadata filtering
**Describe the bug**

I am doing a POC using langcgain4j and spring-boot for our upcoming project. Basically, our requirement is to read one or more documents uploaded by a user and chat or ask questions based on this context. There are multiple users uploaded documents and the answer must be specific to the user context or based on his documents.

So I added the metadata information **userId** when creating the embeddings. I am using the chromadb and I verified with Python that this is working, metadata is successfully added.

The problem is when I am filtering, I am not getting the expected result. When I deep drive into the code I found the filter in EmbeddingSearchRequest(the below function) passing to the ChromaEmbeddingStore.

`default EmbeddingSearchResult<Embedded> search(EmbeddingSearchRequest request) {
        List<EmbeddingMatch<Embedded>> matches = this.findRelevant(request.queryEmbedding(), request.maxResults(), request.minScore());
        return new EmbeddingSearchResult(matches);
    }`

My code: https://github.com/anoopjohn02/spring-ai-langchain4j/blob/dc3027d7ca75c6231e5991c274a97c51715dfdef/src/main/java/com/anoop/ai/config/AIConfig.java#L106

**Log and Stack trace**
NA

**To Reproduce**
Run the app as described in the repo

**Expected behavior**
The dynamic filter should filter doc based on userId

**Please complete the following information:**
- LangChain4j version: 0.32.0
- LLM(s) used: OpenAI
- Java version: 17
- Spring Boot version (if applicable): 3.2.4

**Additional context**
Look at the code in https://github.com/anoopjohn02/spring-ai-langchain4j and let me know my AI configuration is correct.

### PR DESCRIPTION
**Pull request by Heezer**
## Issue

Closes #1505

## Change

Chroma metadata filtering is supported now

## General checklist

- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [X] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features) - https://github.com/langchain4j/langchain4j-examples/pull/99
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for adding new model integration

- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for changing existing embedding store integration

- [X] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

File: langchain4j-chroma/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaMetadataFilterMapper.java [hunk 1]
```diff
@@ -22,8 +22,11 @@ class ChromaMetadataFilterMapper {
         // no instance possible
     }
 
+    // TODO add the tests
```
Comment (langchain4j) [inline langchain4j-chroma/src/main/java/dev/langchain4j/store/embedding/chroma/ChromaMetadataFilterMapper.java]:
```suggestion
```

File: langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java [hunk 2]
```diff
@@ -36,4 +46,183 @@ protected EmbeddingStore<TextSegment> embeddingStore() {
...
+        Filter metadataFilter,
+        List<Metadata> matchingMetadatas,
+        List<Metadata> notMatchingMetadatas
+    ) {}
+
+    // Chroma filters by *not* as following:
+    // If you filter by "key" not equals "a", then in fact all items with "key" != "a" value are returned, but no items
+    // without "key" metadata!
+    // Therefore, all default *not* tests coming from parent class have to be rewritten here.
+    protected static Stream<Arguments> should_filter_by_metadata_not() {
```
Comment (langchain4j) [inline langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java:62]: This method seems to return only a subset of the original test cases, is it on purpose?

File: langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java [hunk 3]
```diff
@@ -36,4 +46,183 @@ protected EmbeddingStore<TextSegment> embeddingStore() {
     protected EmbeddingModel embeddingModel() {
         return embeddingModel;
     }
+
+    @Override
+    @Ignore("Chroma cannot filter by greater and less than of alphanumeric metadata, only int and float are supported")
```
Comment (langchain4j) [inline langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java:51]: Please update `docs/docs/integrations/embedding-stores/index.md` and add a comment with details of Chroma limitations

File: langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java [hunk 4]
```diff
@@ -36,4 +46,183 @@ protected EmbeddingStore<TextSegment> embeddingStore() {
...
+    protected void should_filter_by_greater_and_less_than_alphanumeric_metadata(
+        Filter metadataFilter,
+        List<Metadata> matchingMetadatas,
+        List<Metadata> notMatchingMetadatas
+    ) {}
+
+    // Chroma filters by *not* as following:
```
Comment (langchain4j) [inline langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java:58]: Please update `docs/docs/integrations/embedding-stores/index.md` and add a comment with details of Chroma limitations

**Review by ghost — COMMENTED** [review]
@Heezer thanks a lot, good job!

File: langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java (hunk 2 above)
Comment (Heezer) [reply langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java:62]: yes, on purpose, I've described the reason in comment above

File: langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java (hunk 3 above)
Comment (Heezer) [reply langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java:51]: done

File: langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java (hunk 4 above)
Comment (Heezer) [reply langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java:58]: done

File: langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java (hunk 2 above)
Comment (langchain4j) [reply langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java:62]: The comments says "have to be rewritten", in my understanding it means that this test should have the same number of test cases, but the expected results rewritten. There are 31 test cases in the original test but only 11 here...

Comment (Heezer) [reply langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java:62]: yes, I haven't rewritten all of them, just a random set of them, in my opinion - it should be more that enough to cover the main functionality 😄 how do you see that?

File: langchain4j-core/src/test/java/dev/langchain4j/store/embedding/EmbeddingStoreWithFilteringIT.java [hunk 5]
```diff
@@ -1238,435 +1157,399 @@ void should_filter_by_metadata_not(Filter metadataFilter,
...
+            )
+            .add(
+                Arguments.of(
+                    metadataKey("key").isNotEqualTo(TEST_UUID),
+                    asList(new Metadata().put("key", UUID.randomUUID()), new Metadata().put("key2", TEST_UUID)),
+                    asList(
+                        new Metadata().put("key", TEST_UUID),
+                        new Metadata().put("key", TEST_UUID).put("key2", UUID.randomUUID())
+                    ),
+                    false
```
Comment (langchain4j) [inline langchain4j-core/src/test/java/dev/langchain4j/store/embedding/EmbeddingStoreWithFilteringIT.java]:
```suggestion
                    true
```

**Review by ghost — DISMISSED** [review]
@Heezer thanks a lot, good job!

File: langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java (hunk 2 above)
Comment (langchain4j) [reply langchain4j-chroma/src/test/java/dev/langchain4j/store/embedding/chroma/ChromaEmbeddingStoreIT.java:62]: Fair enough :)

**Pull request merged**